aten/src/THC/THCTensorRandom.cuh

```diff
@@ -140,8 +140,16 @@ __device__ int binarySearchForMultinomial(T* dist,
 
   if (start == size) {
     // No probability mass or precision problems; just return the
-    // first element
-    start = 0;
+    // first non-zero element by setting start to size-1 here,
+    // the code below will move it to the last non-zero probability
+    // this actually can happen when the random number is 1
+    // (github pytorch issue #4858).
+    if (size > 0) {
```

Comment:
We should probably check that numCategories is not 0 in `THCTensor_(multinomial)` instead of here.

Reply:
This would be an additional argcheck in `aten/src/THC/generic/THCTensorRandom.cu`, right?
Currently, it will error out around 0-tensors not having many dimensions, but I'm not sure that that is reliable enough in the long-term.

Reply:
Yes. The problem I was thinking about was with empty tensors that contain no elements but have dimension 1:
```
In [10]: import torch

In [11]: x = torch.Tensor().cuda()

In [12]: x.dim()
Out[12]: 1

In [13]: torch.multinomial(x, 1)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-13-6732e20c3568> in <module>()
----> 1 torch.multinomial(x, 1)

RuntimeError: invalid argument 2: prob_dist must be 1 or 2 dim at /private/home/rzou/p
ytorch/pytorch/aten/src/THC/generic/THCTensorRandom.cu:142
```
But I guess that isn't a problem (this error message should be fixed but that can happen another time). At any rate the size check doesn't need to happen inside the kernel.

Reply:
I'll take it out even if it should not have much impact (because it's only in the rare path).

When feeding a size `(0,)` Tensor `(torch.cuda.FloatTensor([0])[1:]`), I still get the above message before I even take the size of the Tensor. I could imagine that something between python and ATen takes away 0-size dimensions because the function are historically unable to cope with empty dimensions or so, but I don't actually know. 

Best regards

Thomas